data/scripts.js

```diff
@@ -1028,6 +1701,21 @@ let BattleScripts = {
 	targetTypeChoices(targetType) {
 		return CHOOSABLE_TARGETS.has(targetType);
 	},
+
+	reduceMoveResults(states) {
+		return states.reduce((total, val) => {
+			if (typeof total === 'number') {
+				if (typeof val === 'number') {
+					return total + val;
+				} else {
+					return total;
+				}
+			} else if (typeof val === 'number' || (val === true || val === null) && !total || total === undefined) {
+				return val;
+			}
+			return total;
```

Comment:
I'm also not a huge fan of this reducer.

Here's how I'd do it:

First, don't privilege either side of a reducer function. It doesn't matter which side is the total because if your reducer isn't symmetric, it's probably a bug (or if it isn't, it's still less readable).

```js
combineResults(left, right) {
	const NOT_FAILURE = ''; // local so it doesn't need `this` binding
	if (left === NOT_FAILURE) return right;
	if (right === NOT_FAILURE) return left;

	if (typeof left === 'number' && typeof right === 'number') return left + right;
	if (typeof left === 'number') return left;
	if (typeof right === 'number') return right;

	if (left === null || right === null) return null;
	return false;
},
```

Notice that `combineResults` itself _is_ a reducer, rather than _running_ the reducer.

This is so you don't actually need to use `reduce` on single combinations:

```js
// old:
let didAnything = this.reduceMoveResults(damage);
// new:
let didAnything = damage.reduce(this.combineResults);

// old:
didSomething = this.reduceMoveResults([didSomething, false]);
// new:
didSomething = this.combineResults(didSomething, false);
```

Reply:
Is there a way to do this using generics? I tried using
`combineResults?: <T extends number | boolean | null | '' | undefined, U extends number | boolean | null | '' | undefined>(this: Battle, left: T, right: U) => T | U`
```
combineResults(left, right) {
	const NOT_FAILURE = '';
	if (typeof left === 'number' && typeof right === 'number') {
		return left + right;
	} else if (typeof left === 'number') {
		return left;
	} else if (typeof right === 'number') {
		return right;
	} else if (left === null || right === null) {
		return null;
	} else if (left === false || right === false) {
		return false;
	} else if (left === NOT_FAILURE || right === NOT_FAILURE) {
		return NOT_FAILURE;
	} else {
		return undefined;
	}
},
```
in an attempt to get the function to play nice with contexts where `null` and `NOT_FAILURE` aren't relevant, but while this seems right to me, I get errors like
```
Type '<T extends number | boolean | "" | null | undefined, U extends number | boolean | "" | null | undefined>(this: Battle, left: T, right: U) => number | false | "" | null | undefined' is not assignable to type '<T extends number | boolean | "" | null | undefined, U extends number | boolean | "" | null | undefined>(this: Battle, left: T, right: U) => T | U'.
  Type 'number | false | "" | null | undefined' is not assignable to type 'T | U'.
    Type 'undefined' is not assignable to type 'T | U'.
```
 and
`Types T and U have no overlap.`
Is there any hope here, or should I give up on generics?

Reply:
`T` and `U` should be the same type.

`combineResults?: <T extends number | boolean | null | '' | undefined, T extends number | boolean | null | '' | undefined>(this: Battle, left: T, right: T) => T`

And then just recast or ts-ignore all the returns.

Reply:
Also, I believe `undefined` is a success state.

Reply:
Otherwise why is `NOT_FAILURE` not `undefined`?

Reply:
Well, I got Typescript to stop complaining by going with this idea.
```
combineResults(left, right) {
	const NOT_FAILURE = 'string';
	let resultsPriorities = ['undefined', NOT_FAILURE, 'null', 'false', 'number'];
	if (resultsPriorities.indexOf(typeof left) > resultsPriorities.indexOf(typeof right)) {
		return left;
	} else {
		return right;
	}
},
```
However, this entire approach of using `didSomething = this.combineResults(didSomething, hitResults)` instead of just `didSomething = didSomething || hitResult` has been breaking more tests than they've fixed, even though it _should_ be a better way.

Reply:
Well, your `combineResults` implementation seems wrong to me, which might be part of it. :p

I guess I'll look at your code at some point.

Reply:
Yeah, it was wrong. There's just 16 failing tests now!

Reply:
`undefined` isn't a success condition here. Or, it wasn't, but now that you mention it, I should probably use `NOT_FAILURE` as the default starting value instead of `undefined` since `undefined` _is_ a success condition sometimes.

Reply:
Oh, it can't. Because `NOT_FAILURE` means "end this chain", while `undefined` means "keep going".